phys2denoise.py

```diff
@@ -0,0 +1,234 @@
+
+The project is under development.
+
+Copyright 2020, The physiopy community.
+Please scroll to bottom to read full license.
+
+"""
+
+import datetime
+import logging
+import os
+import sys
+from inspect import signature, _empty
+
+import numpy as np
+import pandas as pd
+
+from phys2denoise import _version
+from phys2denoise.cli.run import _get_parser
+from phys2denoise.metrics.cardiac import crf
+from phys2denoise.metrics.chest_belt import rpv, rv, rvt, rrf
+from phys2denoise.metrics.retroicor import compute_retroicor_regressors
+
+from . import __version__
+from .due import due, Doi
+
+LGR = logging.getLogger(__name__)
+
+
+def select_input_args(metric, metric_args):
+    """
+    Retrieve required args for metric from a dictionary of possible arguments.
+
+    This function checks what parameters are accepted by a metric.
+    Then, for each parameter, check if the user provided it or not.
+    If they did not, but the parameter is required, throw an error -
+    unless it's "physio", reserved name for the timeseries input to a metric.
+    Otherwise, use the default.
+
+    Parameters
+    ----------
+    metric : function
+        Metric function to retrieve arguments for
+    metric_args : dict
+        Dictionary containing all arguments for all functions requested by the
+        user
+
+    Returns
+    -------
+    args : dict
+        Arguments to provide as input to metric
+
+    Raises
+    ------
+    ValueError
+        If a required argument is missing
+
+    """
+    args = {}
+
+    # Check the parameters required by the metric and given by the user (see docstring)
+    for param in signature(metric).parameters.values():
+        if param.name not in metric_args:
+            if param.default == _empty and param.name != 'physio':
+                raise ValueError(f'Missing parameter {param} required '
+                                 f'to run {metric}')
+            else:
+                args[param.name] = param.default
+        else:
+            args[param.name] = metric_args[param.name]
+
+    return args
+
+
+def print_metric_call(metric, args):
+    """
+    Log a message to describe how a metric is being called.
+
+    Parameters
+    ----------
+    metric : function
+        Metric function that is being called
+    args : dict
+        Dictionary containing all arguments that are used to parametrise metric
+
+    Notes
+    -----
+    Outcome
+        An info-level message for the logger.
+    """
+    msg = f'The {metric} regressor will be computed using the following parameters:'
+
+    for arg in args:
+        msg = f'{msg}\n    {arg} = {args[arg]}'
+
+    msg = f'{msg}\n'
+
+    LGR.info(msg)
+
+
+@due.dcite(
+     Doi(''),
+     path='phys2denoise',
+     description='Creation of regressors for physiological denoising',
+     version=__version__,
+     cite_module=True)
+def phys2denoise(filename, metric_args, outdir='.',
```

Comment:
I think I'd change `metric_args` to `**kwargs` at the end of the signature. WDYT?

Reply:
`kwargs` would be a dictionary, yes. The way it would be called would be different though. In your case you might do

```python
phys2denoise(filename, metric_args={'a': 1, 'b': 2})
```

but with kwargs you'd do

```python
phys2denoise(filename, a=1, b=2)
```

In either case, documenting those extra arguments is going to be difficult, but I think kwargs is more Pythonic.

The place where I really don't want to see kwargs is in the metric function calls. I'd prefer that they be well-defined, with limited arguments. As long as the keyword arguments to the main workflow function are filtered before they're fed into the metric functions (as you already have planned), I think it's totally fine to have kwargs in the main workflow.

Reply:
Ok, seems right. @ineschh is going to kill me. Sorry @ineschh, I didn't know about this!